Allow and/or blocks to have 0 or 1 inputs

ewpatton opened the pull request:
Change-Id: Ic04587ca376f7ed1bb582d8c273b6f16833ad13d

ewpatton commented:
@mit-cml/core-reviewers This PR now reflects the version I sent out this morning with the tweak to the `+` operator suggested by @halatmit

BeksOmega commented on appinventor/blocklyeditor/src/blocks/logic.js:
```diff
@@ -203,13 +195,34 @@ Blockly.Blocks['logic_operation'] = {
+    if (this.itemCount_ === 0) {
+      this.addEmptyInput();
+    }
+    // NOTE(ewp): Blockly doesn't trigger the validation function when the field is set during
+    // load, so we override setValue here to make sure that the additional and/or labels (if
+    // present) match the dropdown's value.
```
If you mean load as in loading from XML it actually does in the latest version. So you might want to add a TODO to remove this after the update =)

ewpatton replied:
Thanks for the info. Some day it would be nice if we could have the validation/update logic be separated but I expect that's unlikely to happen any time soon. In this particular case, the function is idempotent so it shouldn't matter, but less code executed is better.

ellelili2025 commented:
logic or doesn't look quite right yet. Identifier is missing after I removed all the inputs in the mutators and there are still two inputs showing on the block.
<img width="760" alt="Screen Shot 2020-07-08 at 7 27 36 PM" src="https://user-images.githubusercontent.com/58433071/86980156-1e1dc280-c151-11ea-92ec-846466d2bb70.png">

ewpatton commented:
@ellelili2025 I've addressed your feedback in the following way:

1. This was due to my introduction of a helper method on logic_operator that I didn't also define on logic_or. This should be fixed in the first commit.

2. Implementing this wasn't part of my original goal for this PR, but I've gone ahead and added it.

3. For bitwise-and the identity is -1 (the two's complement of which is all 1s in binary) and for ior/xor it is 0.

ellelili2025 commented:
Logic or looks fine now. But min with 0 input gives me 9223372036854775807 and max gives me -9223372036854775807.

ewpatton commented:
Hmm. When I use Do It I get +Infinity and -Infinity, which is what I would expect.
